src/NuGet.Clients/NuGet.CommandLine/Commands/ClearCacheCommand.cs

```diff
@@ -0,0 +1,140 @@
+            var localAppDataFolderPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            var httpCacheFolderPath = Path.Combine(localAppDataFolderPath, "NuGet", "v3-cache");
+            if (!string.IsNullOrEmpty(httpCacheFolderPath))
+            {
+                Console.WriteLine(
+                    LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_ClearingNuGetHttpCache)),
+                    httpCacheFolderPath);
+
+                var result = ClearCacheDirectory(httpCacheFolderPath);
+                if (commandResult == 0)
+                {
+                    commandResult = result;
+                }
+            }
+
+            if (ClearGlobalPackages)
+            {
+                // also clear the global NuGet packages cache
+                var globalPackagesFolderPath = SettingsUtility.GetGlobalPackagesFolder(Settings);
+
+                Console.WriteLine(
+                    LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_ClearingNuGetGlobalPackagesCache)),
+                    globalPackagesFolderPath);
+
+                var result = ClearCacheDirectory(globalPackagesFolderPath);
+                if (commandResult == 0)
+                {
+                    commandResult = result;
+                }
+            }
+
+            if (commandResult != 0)
+            {
+                throw new CommandLineException(LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_CacheClearFailed)));
+            }
+
+            return Task.FromResult(commandResult);
+        }
+
+        private int ClearCacheDirectory(string folderPath)
+        {
+            // Calling DeleteRecursive rather than Directory.Delete(..., recursive: true)
+            // due to an infrequent exception which can be thrown from that API
+            var failedDeletes = new List<string>();
+            DeleteRecursive(folderPath, failedDeletes);
+
+            if (failedDeletes.Any())
+            {
+                Console.WriteWarning(
+                    LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_CachePartiallyCleared)));
+
+                foreach (var failedDelete in failedDeletes.OrderBy(f => f, StringComparer.OrdinalIgnoreCase))
+                {
+                    Console.WriteWarning(
+                        LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_FailedToDeletePath)),
+                        failedDelete);
+                }
+                return 1;
+            }
+            else
+            {
+                Console.WriteLine(LocalizedResourceManager.GetString(nameof(NuGetResources.ClearCacheCommand_CacheCleared)));
+                return 0;
+            }
+        }
+
+        private static void DeleteRecursive(string deletePath, List<string> failedDeletes)
+        {
+            if (!Directory.Exists(deletePath))
```

Comment:
Contrary to what's documented on MSDN and what one might expect, `Directory.Delete(folderPath, recursive: true);` does NOT delete any files in the directory tree. Therefore, the recursive method to remove any readonly flags will also call into `File.Delete` for each file.

Now also [explicitly checking](https://github.com/NuGet/NuGet.Client/commit/c9e62da0e341331d9084b6d6ecc8f27a8bc64452#diff-16d0285c2ebe50837c5df60fd56b56a1R148) for the presence of `FileAttributes.ReparsePoint` on directories.

Please review commits https://github.com/NuGet/NuGet.Client/commit/c9e62da0e341331d9084b6d6ecc8f27a8bc64452 and https://github.com/NuGet/NuGet.Client/commit/ccd4ea684fb978a2f8deeb8ef38614e460dbe11f